### PR TITLE
Refactor simplify preprocessing

### PR DESCRIPTION
**Pull request by florian-hoenicke**
changes:
- preprocessing is unified and not app-specific anymore. A modality is always preprocessed in the same way
- everything that needs to be encoded is put into cc level after the preprocessor, i.e., images are on cc level and video frames as well. Also, sentences are on cc level
- indexer only cares about cc level and does not make a distinction anymore between different levels
- ocr computation simplified
- ci bug fixed - before some executors were not rebuilt 
- fine-tuning deactivated - we have no case at the moment where the improvement justifies the lack of generalization

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -18,3 +18,4 @@ boto3
 dask[distributed]==2022.2.0
 jina-hubble-sdk>=0.15.5
 elasticsearch==8.0.1
+pydantic==1.9.2
```
Comment (florian-hoenicke) [inline requirements.txt:21]: This was required for the qdrant executor. Otherwise the build takes very long and is stale at some point

**Review by joschkabraun — CHANGES_REQUESTED** [review]
Nice refactoring!
Left some comments

File: deployment/bff/app/v1/routers/im_txt2im_txt.py [hunk 2]
```diff
@@ -76,8 +72,7 @@ def search(data: NowTextImageSearchRequestModel):
         parameters={
             'limit': data.limit,
             'filter': filter_query,
-            'access_paths': '@c',
-            'traversal_paths': '@c',
+            'access_paths': '@cc',
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply deployment/bff/app/v1/routers/im_txt2im_txt.py:75]: the external clip encoder still requires the path. If we deploy that one with access_paths cc then it is solved

File: now/executor/preprocessor/Dockerfile [hunk 3]
```diff
@@ -6,13 +6,15 @@ RUN apt-get update && apt-get install -y git &&\
     apt-get -y install build-essential libgomp1 libgl1-mesa-glx libglib2.0-0 libsm6 libxext6 libxrender-dev && \
     rm -rf /var/cache/apt/*
 
+    # setup the workspace
+COPY . /workdir/
+WORKDIR /workdir
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/executor/preprocessor/Dockerfile:11]: Whenever the files in the work dir are changed, the cache is invalidated

File: now/app/base/app.py [hunk 4]
```diff
@@ -30,21 +36,21 @@
...
~        """
~        Set to True if this app is enabled for the end user.
~        """
~        raise NotImplementedError()
~
~    @property
~    def description(self) -> str:
~        """
~        Short description of the app.
~        """
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/app/base/app.py]: thanks done

File: now/app/base/preprocess.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
+import io
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/app/base/preprocess.py:1]: Nice finding. I changed it to 
```            
            has_preprocessing_changed:
            - 'now/app/base/**'
            - 'now/common/preprocess.py'
            - 'now/executor/preprocessor/**'
            - 'now/executor/abstract/auth/**'
```

File: tests/common/test_preprocess.py [hunk 6]
```diff
@@ -1,50 +0,0 @@
...
-
-
-def test_preprocess_short_da_no_split():
-    da = DocumentArray(
-        [Document(chunks=[Document(text='test. test', modality='text')])]
-    )
-    result = preprocess_text(da, split_by_sentences=False)
-    assert len(result) == 1
-    assert len(result[0].chunks) == 1
-    assert len(result[0].chunks[0].chunks) == 0
```
Comment (violenil) [inline tests/common/test_preprocess.py:50]: Are these text preprocessing tests still present somewhere else, ie. is this functionality (like splitting) tested elsewhere?

File: tests/executor/preprocessor/test_unit.py [hunk 7]
```diff
@@ -60,10 +58,10 @@ def test_s3_video():
...
-def test_s3_text():
+def test_text():
     da_search = DocumentArray([Document(text='test')])
     preprocessor = NOWPreprocessor(Apps.TEXT_TO_VIDEO)
-    res_search = preprocessor.search(da_search, parameters={})
+    res_search = preprocessor.preprocess(da_search, parameters={})
     assert len(res_search) == 1
     assert len(res_search[0].chunks) == 1
-    assert res_search[0].chunks[0].text == 'test'
+    assert res_search[0].chunks[0].chunks[0].text == 'test'
```
Comment (violenil) [inline tests/executor/preprocessor/test_unit.py:67]: Might be nice to add a case here with a document where text is a bit longer (2 or 3 sentences) and make sure all sentences are now chunks.

**Review by violenil — COMMENTED** [review]
I love this refactoring! Good work 🚀 left some questions/comments.

File: now/app/base/preprocess.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+from now.constants import Modalities
+from now.data_loading.convert_datasets_to_jpeg import to_thumbnail_jpg
+
+NUM_FRAMES_SAMPLED = 3
+
+
+def preprocess_text(
+    d: Document,
+) -> Document:
+    """If necessary, loads text for all documents. If asked for, splits documents by sentences.
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/app/base/preprocess.py:16]: Thanks for noticing. I will update the docstring

File: now/executor/abstract/base_indexer/base_indexer.py [hunk 9]
```diff
@@ -107,26 +108,24 @@ def list(self, parameters: dict = {}, **kwargs):
...
-                    break
-                if d.parent_id in parent_ids:
-                    continue
-                parent_ids.add(d.parent_id)
-                docs.append(d)
-        else:
-            docs = self.document_list[offset : offset + limit]
+
+        docs = DocumentArray()
+        chunks_size = int(parameters.get('chunks_size', 3))
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/executor/abstract/base_indexer/base_indexer.py]: Very good finding! chunks_size was just some bad hack I did in the past. Each modality might have a variable number of chunks. Therefore, a constant number does not make sense

File: now/executor/abstract/base_indexer/base_indexer.py [hunk 10]
```diff
@@ -145,8 +144,8 @@ def index_endpoint(
         :param docs: the Documents to index
         :param parameters: dictionary with options for indexing
         """
-        traversal_paths = parameters.get('traversal_paths', self.traversal_paths)
-        flat_docs = docs[traversal_paths]
+        flat_docs = docs[ACCESS_PATHS]
+        # TODO please remove theses checkf for empty docs
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/executor/abstract/base_indexer/base_indexer.py:152]: thanks, I don't know what I was writing :)

File: now/executor/autocomplete/executor.py [hunk 11]
```diff
@@ -49,7 +48,8 @@ def update_save_words(self, word):
     def search_update(
         self, docs: Optional[DocumentArray] = None, parameters: dict = {}, **kwargs
     ):
-        flat_docs = docs[self.search_traversal_paths]
+        # TODO needs to be also on @cc path after preprocessing for simplification
+        flat_docs = docs['@r']
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/executor/autocomplete/executor.py:52]: True thanks

File: now/executor/autocomplete/executor.py [hunk 12]
```diff
@@ -69,12 +69,11 @@ def search_update(
...
         self, docs: Optional[DocumentArray] = None, parameters: dict = {}, **kwargs
     ):
-        flat_docs = None if not docs else docs[self.search_traversal_paths]
-        if flat_docs:
-            for doc in flat_docs:
-                doc.tags['suggestions'] = self.flatten_list(
-                    self.auto_complete.search(doc.text, max_cost=3, size=5)
-                )
+        # TODO needs to be also on @cc path after preprocessing for simplification
+        for doc in docs['@r']:
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/executor/autocomplete/executor.py:73]: thanks

File: now/executor/indexer/elastic/elastic_indexer.py [hunk 13]
```diff
@@ -70,30 +67,18 @@ def _generate_es_mapping(self, dims: List[int]) -> Dict:
...
-                es_mapping['properties'][f'chunk_{i}'] = {
-                    'properties': {
-                        f'embedding': {
-                            'type': 'dense_vector',
-                            'dims': dim,
-                            'similarity': self.metric,
-                            'index': 'true',
-                        }
+
+        for i, dim in enumerate(self.dims):
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply now/executor/indexer/elastic/elastic_indexer.py:71]: Thanks, yes. I was just trying to get it passed.

File: tests/executor/preprocessor/test_unit.py (hunk 7 above)
Comment (florian-hoenicke) [reply tests/executor/preprocessor/test_unit.py:67]: Nice, done

File: tests/common/test_preprocess.py (hunk 6 above)
Comment (florian-hoenicke) [reply tests/common/test_preprocess.py:50]: Good catch - added that case

**Pull request merged**